- Description: cio: Dasds not available after set online.
  Symptom:     Certain types of dasds can't be used when attempting to set them
               online.
  Problem:     Some dasds support SensePGID, but not SetPGID. The common I/O
               layer nevertheless tries SetPGID on every path and eventually
               switches off every path, making the device not available.
  Solution:    If SetPGID returns with command reject both for multi path and
               single path operation, consider path grouping not supported and
               just set the device online.
  Problem-ID:  -
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/cio/device_pgid.c	23 Apr 2004 14:40:37 -0000	1.21.2.1
+++ linux-2.5/drivers/s390/cio/device_pgid.c	14 May 2004 11:27:49 -0000	1.21.2.2
@@ -338,6 +338,10 @@
 		 * One of those strange devices which claim to be able
 		 * to do multipathing but not for Set Path Group ID.
 		 */
+		if (cdev->private->flags.pgid_single) {
+			ccw_device_verify_done(cdev, -EOPNOTSUPP);
+			break;
+		}
 		cdev->private->flags.pgid_single = 1;
 		/* fall through. */
 	case -EAGAIN:		/* Try again. */

--- linux-2.5/drivers/s390/cio/device_fsm.c	23 Apr 2004 12:57:40 -0000	1.85.2.2
+++ linux-2.5/drivers/s390/cio/device_fsm.c	14 May 2004 11:27:49 -0000	1.85.2.3
@@ -473,6 +473,8 @@
 {
 	cdev->private->flags.doverify = 0;
 	switch (err) {
+	case -EOPNOTSUPP: /* path grouping not supported, just set online. */
+		cdev->private->options.pgroup = 0;
 	case 0:
 		ccw_device_done(cdev, DEV_STATE_ONLINE);
 		break;

